File: python/bKap5_6_OutOfSample.py
"""
bKap5_6_OutOfSample.py
Python port of bKap5_6_OutOfSample.m.

PLS dimensionality reduction (out-of-sample). At each origin a PLS + regression
is re-estimated on a rolling/expanding window and used to forecast inflation one
step ahead, compared against the historical-mean benchmark.

Run from the repo root:  python python/bKap5_6_OutOfSample.py
"""

import numpy as np
import pandas as pd

import models
import util

# =========================================================================
#  Settings  (mirror the MATLAB header)
# =========================================================================
DATA_PATH    = "./DATA/Liedtke/US/aggregated.csv"

iTimeLag    = 1      # additional predictive lag
iNumComp    = 3      # number of PLS components to retain
iTransformX = 2      # 0 = none, 1 = centre, 2 = z-standardise
iNumIn      = 240    # minimum in-sample observations before forecasting
lRoll       = False  # False = expanding window, True = rolling window

# =========================================================================
#  Load data
# =========================================================================
tData = pd.read_csv(DATA_PATH)
dtDates = pd.to_datetime(tData.iloc[:, 0])
mX = tData.iloc[:, 2:].to_numpy(float)
vY = tData.iloc[:, 1].to_numpy(float)

iNumObs, iNumPredictors = mX.shape

# Lag the predictors so they are known one period ahead of the target
mXlag = np.full_like(mX, np.nan)
mXlag[iTimeLag:, :] = mX[:-iTimeLag, :]

# Remove missing values row-wise (PLS does not allow missing values)
lIsNaN = np.isnan(vY) | np.isnan(mXlag).any(axis=1)
keep = ~lIsNaN
vY = vY[keep]
mXlag = mXlag[keep, :]
dtDates = dtDates[keep].reset_index(drop=True)

iNumObs = len(vY)
iNumComp = min(iNumComp, iNumPredictors)

# =========================================================================
#  Out-of-sample analysis (see bKap5_5_OutOfSamplePCA for the index mapping)
# =========================================================================
vYhat = np.full(iNumObs, np.nan)     # PLS forecast
vYroll = np.full(iNumObs, np.nan)    # historical-mean benchmark

for t in range(iNumIn - 1, iNumObs):
    if lRoll:
        idx_in = np.arange(t - iNumIn + 1, t)
    else:
        idx_in = np.arange(0, t)

    mXin = mXlag[idx_in, :]
    vXout = mXlag[t, :]
    vYin = vY[idx_in]

    # Estimate PLS on the in-sample data; project the OOS row with the
    # in-sample mean/std and weights (no look-ahead).
    rModel = models.est_pls(vYin, mXin, num_comp=iNumComp, transform_x=iTransformX)
    mScoresIn = rModel["scores"]
    mScoresOut = models.project_data(vXout, rModel["weights_x"],
                                     rModel["mean_x"], rModel["std_x"])

    # Regress inflation on the in-sample PLS scores (with constant)
    Xd = np.column_stack([np.ones(len(vYin)), mScoresIn])
    beta, *_ = np.linalg.lstsq(Xd, vYin, rcond=None)
    vYhat[t] = np.concatenate([[1.0], mScoresOut.ravel()]) @ beta

    vYroll[t] = vYin.mean()

# =========================================================================
#  Performance evaluation
# =========================================================================
oos = util.evaluate_oos(vY, vYroll, vYhat)
sWindow = "rolling" if lRoll else "expanding"
print("One-step-ahead OOS inflation forecast (PLS)")
print(f"  Comps = {iNumComp} | window = {sWindow} (min in-sample {iNumIn}) | "
      f"forecasts = {int(np.sum(~np.isnan(vYhat)))}\n")

# Table (mirrors cTable_OOS): OOS R2 | CW (p) | OOS R2 CT | CW (p) CT, in %
mResults = np.round([oos["R2OOS"] * 100, oos["CWp"] * 100,
                     oos["R2OOS_CT"] * 100, oos["CWp_CT"] * 100], 2)
header = ["OOS R2", "CW (p)", "OOS R2 CT", "CW (p) CT"]
print("  ".join(f"{h:>10s}" for h in header))
print("  ".join(f"{v:>10.2f}" for v in mResults))

File: bKap5_6_OutOfSample.m
% Script for applying PLS dimensionality reduction to the Liedtke macro panel
% (out-of-sample). At each origin a PLS + regression is re-estimated on a
% rolling/expanding window and used to forecast inflation one step ahead. The
% forecast is compared against the historical-mean benchmark.

% Clear console
clear; clc; close all;

% Set paths
sOldPath = path;
sCountry = fCfg('COUNTRY', 'US');
sDataPath = ['./DATA/Liedtke/', sCountry, '/'];
sResultsPath = './RESULTS/GWZ/';
addpath('./Utils/');

%% Settings
iTimeLag    = 1;     % additional predictive lag (predictors are already
                     % reporting-lag aligned in DATA/Liedtke/aggregate.py)
iNumComp    = 3;     % number of PLS components to retain
iTransformX = 2;     % preprocessing: 0 = none, 1 = centre, 2 = z-standardise
iNumIn      = 240;   % minimum in-sample observations before forecasting
lRoll       = fCfg('ROLLING', false); % false = expanding window, true = rolling window

%% Load data
% Macro panel produced by DATA/Liedtke/aggregate.py.
% CSV layout: col 1 = observation_date, col 2 = target (inflation),
%             col 3..end = predictors (already lag-aligned in Python).
tData     = readtable([sDataPath, 'aggregated.csv']);
cAllNames = tData.Properties.VariableNames;

% Parse dates (first column)
dtDates = tData{:,1};
if ~isdatetime(dtDates)
    dtDates = datetime(dtDates);
end

% Target (col 2) and predictors (col 3..end)
mX          = tData{:,3:end};          % macro predictors
vY          = tData{:,2};              % inflation target
cXnamesM    = cAllNames(3:end)';       % predictor names as an Nx1 cell

%% Data preprocessing
% Determine dimensions
[iNumObs, iNumAssets]       = size(vY);
[iNumObsP, iNumPredictors]  = size(mX);

% Lag the predictors so that they are known one period ahead of the target
mXlag = [NaN(iTimeLag, iNumPredictors); mX(1:end-iTimeLag,:)];

% Remove missing values row-wise (PLS does not allow missing values)
lIsNaN          = isnan(vY) | any(isnan(mXlag),2);
vY(lIsNaN)      = [];
mXlag(lIsNaN,:) = [];
dtDates(lIsNaN) = [];

% Number of observations
iNumObs  = size(vY,1);
iNumComp = min(iNumComp, iNumPredictors);

%% Out-of-sample analysis
% Initialize memory
vYhat   = NaN(iNumObs,1);     % PLS forecast
vYroll  = NaN(iNumObs,1);     % historical-mean benchmark

% Loop over time
for iIdxT = iNumIn:iNumObs
    % Get time indices
    if lRoll
        vIdxInSample = (iIdxT-iNumIn+1):(iIdxT-1);
    else
        vIdxInSample = 1:(iIdxT-1);
    end

    % Get data
    mXin    = mXlag(vIdxInSample,:);        % t-1
    vXout   = mXlag(iIdxT,:);               % t-1
    vYin    = vY(vIdxInSample,:);           % t

    % Estimate PLS on the in-sample data and project the out-of-sample row
    % using the IN-SAMPLE mean/std and weights (no look-ahead).
    rModel     = fEstPLS(vYin, mXin, "iNumComp", iNumComp, "iTransformX", iTransformX);
    mScoresIn  = rModel.mScores;
    mScoresOut = fProjectData(vXout, rModel.mWeightsX, rModel.vMeanX, rModel.vStdX);

    % Regress inflation on the in-sample PLS scores (with constant)
    mBetaPLS     = [ones(size(mXin,1), 1), mScoresIn] \ vYin;
    vYhat(iIdxT) = [1, mScoresOut] * mBetaPLS;

    % Historical-mean benchmark prediction
    vYroll(iIdxT) = mean(vYin);
end

%% Performance evaluation
[rStatsOOS] = fEvaluatePerformanceOOS(vY, vYroll, vYhat);

if lRoll; sWindow = 'rolling'; else; sWindow = 'expanding'; end
fprintf('One-step-ahead OOS inflation forecast (PLS)\n');
fprintf('  Comps = %d | window = %s (min in-sample %d) | forecasts = %d\n\n', ...
    iNumComp, sWindow, iNumIn, sum(~isnan(vYhat)));

% === Create table
% Merge all results and round to two digits
mResults = round([rStatsOOS.vR2OOS * 100; ...
    rStatsOOS.vCWp * 100; ...
    rStatsOOS.vR2OOSCT * 100; ...
    rStatsOOS.vCWp_CT * 100], 2);

% Make to cell and add column header
cTable_OOS = sprintfc('%.2f', mResults');
cTable_OOS = [{'OOS R2', 'CW (p)', 'OOS R2 CT', 'CW (p) CT'}; cTable_OOS];
disp(cTable_OOS);

% === Save results
% Structured output: <GWZ>/PLS/<country>/oos/<options>/
sCountry = fCountryFromPath(sDataPath);
sOutDir  = fResultDir(sResultsPath, 'PLS', sCountry, 'oos', ...
    sprintf('comp%d_min%d_%s', iNumComp, iNumIn, sWindow));
writecell(cTable_OOS, fullfile(sOutDir, 'results.csv'));
save(fullfile(sOutDir, 'results.mat'), 'cTable_OOS', 'vYhat', 'vYroll', 'rStatsOOS');

% Restore path
path(sOldPath);
